docs: tweak documentation

richb-hanover opened the pull request:
Several changes to documentation to include more links between pages:
- Add links to keywords for more info;
- Tweak description of strings;
- Add literals to Syntax README;
- Change `sort` to say "tuple"

richb-hanover committed:
Add links to more info for keywords;
Tweak description of strings;
Add all literals to Syntax README;
Change `sort` to say "tuple"

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

richb-hanover committed:
Fix link to null.md;
cibot fixed line break;

richb-hanover committed:
Merge branch 'tidy-docs' of https://github.com/richb-hanover/prql into tidy-docs

max-sixty commented on web/book/src/reference/syntax/README.md:
```diff
@@ -8,24 +8,27 @@ A summary of PRQL syntax:
+| `[]`                   | [Arrays](./arrays.md)                                                          | `[1, 4, 3, 4]`                                          |
+| `! && == +`            | [Operations](./operators.md)                                                   | <code>filter a == b + c \|\| d >= e</code>              |
+| `()`                   | [Parentheses](./operators.md#parentheses)                                      | `derive celsius = (fahrenheit - 32) / 1.8`              |
+| `123`,`5e10`,`100_000` | [Numbers](./literals.md#numbers)                                               | `derive { huge = 5e10 }`                                |
```
I think this is the main addition in this table? Good idea

richb-hanover replied:
The major change is inclusion of the remaining items in the Literals page: Numbers, Boolean, Null.

max-sixty replied:
Ah yes, I see, the next couple of lines after `Strings` too.

I'm not really sure `true` / `null` are _syntax_, but don't have a clear view, can see the benefit of having them here, and defer to you.

max-sixty commented on web/book/src/SUMMARY.md:
```diff
@@ -7,6 +7,8 @@
 
   - [Literals](./reference/syntax/literals.md)
+    - [F-strings](./reference/syntax/f-strings.md)
+    - [S-strings](./reference/syntax/s-strings.md)
```
I am +1 on making more of a hierarchy here

...but are these literals? I think technically they're not — they are not constant — PRQL changes them...

WDYT?

(we can revert this and merge, or discuss before merging, as you wish)

richb-hanover replied:
I agree that s-strings and f-strings aren't strictly literals, but the reader will be comforted to see all the "string-like" things lumped together. (I don't think those string-types would be any easier for newcomers to find farther down in the Syntax table of content.) 

Let's commit as-is and see if anyone comments.

max-sixty replied:
I'm hesitant to add something that is wrong... What do others think @PRQL/prql-team ?

Is there a way of adding a bit of hierarchy here while keeping it correct?

richb-hanover replied:
In an [earlier life (wayback machine from March 2023)](https://web.archive.org/web/20230313115935/https://prql-lang.org/book/) we had a "Language Features" section with Strings, F-strings and S-strings (sections 5.8, 5.9, 5.10) grouped together. 

Perhaps the Strings paragraphs could be factored from the Literals page into a separate page, with S- and F- string pages (and R-string page?) following, or as a sub-pages? The "Syntax README" would still have a link to the main Strings page, which would continue to refer to the other formats.

richb-hanover replied:
@max-sixty I don't know if others have responded to this PR/suggestion. What do you think?

max-sixty replied:
Sorry, was out for a couple of days (2nd child!), just back home :)

> In an [earlier life (wayback machine from March 2023)](https://web.archive.org/web/20230313115935/https://prql-lang.org/book/)

lol

---

I think at the moment two options:
- Change `Literals` to `Literals & Strings`, retain the proposed approach of having F & S strings within thta
- Revert to the existing approach of separating `Literals` from F & S strings

Your call on which (or happy if you want to propose an alternative)

richb-hanover replied:
OK. I am working on factoring out Strings with sub-headings. Coming soon to a theater near you - probably later today :-)

max-sixty commented on web/book/src/reference/syntax/keywords.md:
```diff
@@ -88,23 +88,25 @@ definitions: `tracks`, `public.tracks`, `my_database.public.tracks`.
-within std namespace:
+Transforms are normal functions within the `std` namespace, not keywords. That
+is, `std.from` is the same function as `from`. In the example below, the
+resulting query is the same as without the `std.` namespace:
```
+1

max-sixty commented on web/book/src/reference/syntax/literals.md:
```diff
@@ -46,26 +46,8 @@ Strings can contain any escapes defined by
-Dialects interpret escape characters differently, and PRQL doesn't yet account for
-these differences. Please open issues with any difficulties in the current
-implementation.
+derive escapes = "\tXYZ\n \\ "                  # tab (\t), "XYZ", newline (\n), " ", \, " "
```
+1, thanks!

max-sixty commented on web/book/src/reference/syntax/literals.md:
```diff
@@ -82,37 +64,60 @@ select {
-- [S-strings](./s-strings.md) - Insert SQL statements directly into the query.
-  Use when PRQL doesn't have an equivalent facility.
+- [**R-strings**](#r-strings) - Include the raw characters of the string without
+  any form of escaping.
```
Great!

max-sixty commented on web/book/src/reference/syntax/literals.md:
```diff
@@ -167,7 +172,7 @@ derive first_check_in = start + 10days
 ### Examples
 
-Here's a fuller list of date and time examples:
+Here's a larger list of date and time examples:
```
Yes better. 

Your grammar is better than mine!

richb-hanover replied:
Probably not, but I have more time to spend on fussy stuff... :-)

max-sixty reviewed (commented):
Excellent, pending one question re literals...

richb-hanover committed:
Merge branch 'main' into tidy-docs

richb-hanover committed:
Update SQL snapshot

richb-hanover committed:
Merge commit 'ae122df5a55bebbe0b992083fbd2794918f909bb' into tidy-docs

richb-hanover committed:
Merge commit 'ae7831a50f34980914c4af701f50ba12cbbd8141' into tidy-docs

richb-hanover committed:
Factor Strings from Literals into a separate file with sub-pages;
Move S-strings and F-strings below main Strings page;
Create new Strings and R-strings page;

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

richb-hanover committed:
Fix silly typo's

richb-hanover committed:
Merge branch 'tidy-docs' of https://github.com/richb-hanover/prql into tidy-docs

# Conflicts:
#	web/book/src/reference/syntax/strings.md

richb-hanover committed:
Merge branch 'main' into tidy-docs

richb-hanover committed:
Fix link to r-strings.md

richb-hanover committed:
Merge branch 'tidy-docs' of https://github.com/richb-hanover/prql into tidy-docs

max-sixty commented on web/book/src/reference/syntax/s-strings.md:
````diff
@@ -90,7 +92,7 @@ derive {
 ```
 
-Instead, the denominator `{gross_salary}` must be encased in parentheses:
+Instead, the numerator `{gross_salary}` must be encased in parentheses:
````
whoops!

max-sixty commented on web/book/src/reference/syntax/strings.md:
```diff
@@ -15,7 +15,7 @@ derive {
 
 To quote a string containing quote characters, use the "other" type of quote, or
-use the escape character ("\"), or use more quotes.
+use the escape character ("\\"), or use more quotes.
```
```suggestion
use the escape character (`\`), or use more quotes.
```
Is that clearer? I worry that `"\\"` isn't the escape character?

richb-hanover replied:
Nope - the markdown needs to escape the escape character - two `\` are required

max-sixty replied:
This works:

`````diff
diff --git a/web/book/src/reference/syntax/strings.md b/web/book/src/reference/syntax/strings.md
index ce103f80..8d0dd919 100644
--- a/web/book/src/reference/syntax/strings.md
+++ b/web/book/src/reference/syntax/strings.md
@@ -15,7 +15,7 @@ # Strings
 **Quoting and escape characters**
 
 To quote a string containing quote characters, use the "other" type of quote, or
-use the escape character ("\\"), or use more quotes.
+use the escape character (`\`), or use more quotes.
 
 ```prql
 from artists
`````

max-sixty commented on web/book/src/reference/syntax/literals.md:
````diff
@@ -27,92 +27,24 @@ select {
 }
 ```
 
-## Strings
````
Can we put a link to the Strings page at the bottom of the page (e.g. "See also" or similar), since the normal string is a literal?

richb-hanover replied:
Good point. Done

max-sixty replied:
Great, can merge as soon as you push!

richb-hanover committed:
Fix link to strings.md

richb-hanover committed:
Merge branch 'main' into tidy-docs
